dotcom-rendering/src/web/components/DynamicFast.tsx

```diff
@@ -1,58 +1,434 @@
-import type { DCRContainerPalette } from '../../types/front';
+import { css } from '@emotion/react';
+import { until } from '@guardian/source-foundations';
+import type { DCRContainerPalette, DCRGroupedTrails } from '../../types/front';
 import { LI } from './Card/components/LI';
 import { UL } from './Card/components/UL';
 import { FrontCard } from './FrontCard';
 
 type Props = {
-	trails: TrailType[];
+	groupedTrails: DCRGroupedTrails;
 	containerPalette?: DCRContainerPalette;
 	showAge?: boolean;
 };
 
-export const DynamicFast = ({ trails, containerPalette, showAge }: Props) => {
-	const primary = trails[0];
-	const secondary = trails[1];
-	const bigCards = trails.slice(2, 4);
-	const smallCards = trails.slice(4, 10);
+/**
+ * 'Primaries' renders the first row of dynamic/fast
+ * This row will consist of either 1 huge, 1 very big, or 2 very bigs.
+ *
+ * When there is 1 huge or 1 very big, these will be treated the same, and take
+ * up 100% of the horizontal space
+ *
+ * When there are 2 very bigs in the primaries row, they each take 50% of the horizontal space.
+ * Either of the 2 very bigs can 'boosted', in which case the boosted card will receive
+ * 75% of the horizontal space.
+ */
+const Primaries = ({
+	primaries,
+	showAge,
+	containerPalette,
+}: {
+	primaries: [] | [TrailType] | [TrailType, TrailType];
+	showAge?: boolean;
+	containerPalette?: DCRContainerPalette;
+}) => {
+	// Gets the index of the first primary card which is boosted, returns -1 if none are.
+	const boostedIndex = primaries.findIndex((card) => !!card.isBoosted);
 
 	return (
-		<>
-			<UL direction="row" padBottom={true}>
-				<LI padSides={true} percentage="75%">
-					<FrontCard
-						trail={primary}
-						containerPalette={containerPalette}
-						showAge={showAge}
-						//Overrides
-						headlineSize="large"
-						imagePosition="right"
-						imagePositionOnMobile="top"
-						imageSize="large"
-						supportingContent={primary.supportingContent}
-						trailText={primary.trailText}
-					/>
-				</LI>
+		<UL direction="row" padBottom={true}>
+			{primaries.map((card, index) => {
+				// Boosted Layout
+				if (primaries.length > 1 && boostedIndex !== -1) {
```

Comment:
fwiw, this is the section that tripped me up when reading it. But, like other reviewers, I agree that the complexity is pretty inherent to the setup here, and I can't think of a better alternative!

It feels to me like there are two main sources of complexity here when trying to understand this section. One is the number of different layouts that can be generated. The other is that the `isBoosted` property that we get through from tools/Frontend doesn't actually mean that the card will be rendered as boosted (hence having to check against `boostedIndex` here, rather than being able to check against `isBoosted`). Does that sound right?

I think the point is implicit in your comment on ln.34, but it might be helpful to make it a bit more explicit? i.e. something like "A maximum of one card can be boosted in this slice, but this requirement is not enforced in the data we receive, so we are using `boostedIndex` here to determine whether a card should actually be boosted, rather than relying on the `isBoosted` property directly."?

Reply:
This was a good suggestion & from reading I think you understand it all well - I will update this comment to be more explicit now!